go/ai/tools.go

```diff
@@ -29,30 +33,91 @@ const provider = "local"
 // The ToolDefinition has JSON schemas that describe the types.
 // TODO: This should be generic over the function input and output types,
 // and something in the general code should handle the JSON conversion.
-type Tool struct {
-	Definition *ToolDefinition
-	Fn         func(context.Context, map[string]any) (map[string]any, error)
+type Tool[In, Out any] struct {
```

Comment:
It's not immediately clear to me why we need both `Tool` and `toolAction`. Can you explain?

Reply:
The way generics in Go work. I need a type that doesn't have generics to construct and return from `LookupTools`. `toolAction` is it. Otherwise when holding on to a reference to a `Tool[In, Out]` you have strongly typed (with generics) reference.